### PR TITLE
Always select the in-memory group/version as a target when decoding from storage

### LINKED ISSUE #73478: Cross-group watch events from etcd are not handled correctly
**What happened**:
* Set up a HA cluster with 1.9 and 1.10 API servers (occurs on any apiserver skew boundary where a cohabitating resource's storage version changed groups)
* Observed errors when modifying workload objects:
  * when modified via the 1.10 server, saw this error in the 1.9 server: `failed to prepare current and previous objects: v1.Deployment is not suitable for converting to {{"extensions" "__internal"} [{"extensions" ""} {"extensions" ""}]}`
  * when modified via the 1.9 server, saw this error in the 1.10 server: `failed to prepare current and previous objects: v1beta1.Deployment is not suitable for converting to {{"apps" "__internal"} [{"apps" ""} {"apps" ""}]}`
* When this error was encountered, all watchers for the resource in question would be dropped and have to re-establish, and the internal watchers feeding the watch cache would drop and re-establish

**What you expected to happen**:
* Watch events from all supported storage versions are processed correctly

**How to reproduce it (as minimally and precisely as possible)**:
* Start a server from master (for inspectability, store in json). With https://github.com/kubernetes/kubernetes/pull/73477:
    ```
    export STORAGE_MEDIA_TYPE=application/json
    export KUBE_PANIC_WATCH_DECODE_ERROR=false
    hack/local-up-cluster.sh
    ```
* Create a deployment (e.g. kube-dns in kube-system)
* Establish a watch via extensions (kubectl get deployments.extensions --all-namespaces -w)
* Establish a watch via apps (kubectl get deployments.apps --all-namespaces -w)
* Tail the kube-apiserver log
* Change the apps/v1 deployment in etcd to extensions/v1beta1:
    ```
    ETCDCTL_API=3 etcdctl get /registry/deployments/kube-system/kube-dns --print-value-only \
    | \
    kubectl convert -f - --output-version=extensions/v1beta1 -o json \
    | \
    jq '.items[0] | .metadata.namespace = "kube-system"' \
    | \
    ETCDCTL_API=3 etcdctl put /registry/deployments/kube-system/kube-dns
    ```
* Observer client watches drop and server log errors:
```
E0129 10:42:55.911298   76283 watcher.go:268] failed to prepare current and previous objects: v1beta1.Deployment is not suitable for converting to {{"apps" "__internal"} [{"apps" ""} {"apps" ""}]} in scheme "k8s.io/kubernetes/pkg/api/legacyscheme/scheme.go:29"
E0129 10:42:55.911298   76283 watcher.go:268] failed to prepare current and previous objects: v1beta1.Deployment is not suitable for converting to {{"apps" "__internal"} [{"apps" ""} {"apps" ""}]} in scheme "k8s.io/kubernetes/pkg/api/legacyscheme/scheme.go:29"
E0129 10:42:55.911322   76283 watcher.go:268] failed to prepare current and previous objects: v1beta1.Deployment is not suitable for converting to {{"apps" "__internal"} [{"apps" ""} {"apps" ""}]} in scheme "k8s.io/kubernetes/pkg/api/legacyscheme/scheme.go:29"
E0129 10:42:55.911494   76283 watcher.go:268] failed to prepare current and previous objects: v1beta1.Deployment is not suitable for converting to {{"apps" "__internal"} [{"apps" ""} {"apps" ""}]} in scheme "k8s.io/kubernetes/pkg/api/legacyscheme/scheme.go:29"
W0129 10:42:55.914348   76283 reflector.go:303] storage/cacher.go:/deployments: watch of *apps.Deployment ended with: Internal error occurred: v1beta1.Deployment is not suitable for converting to {{"apps" "__internal"} [{"apps" ""} {"apps" ""}]} in scheme "k8s.io/kubernetes/pkg/api/legacyscheme/scheme.go:29"
W0129 10:42:55.914402   76283 reflector.go:303] storage/cacher.go:/deployments: watch of *apps.Deployment ended with: Internal error occurred: v1beta1.Deployment is not suitable for converting to {{"apps" "__internal"} [{"apps" ""} {"apps" ""}]} in scheme "k8s.io/kubernetes/pkg/api/legacyscheme/scheme.go:29"
W0129 10:42:55.914348   76283 reflector.go:303] storage/cacher.go:/deployments: watch of *apps.Deployment ended with: Internal error occurred: v1beta1.Deployment is not suitable for converting to {{"apps" "__internal"} [{"apps" ""} {"apps" ""}]} in scheme "k8s.io/kubernetes/pkg/api/legacyscheme/scheme.go:29"
W0129 10:42:55.914478   76283 reflector.go:303] storage/cacher.go:/deployments: watch of *apps.Deployment ended with: Internal error occurred: v1beta1.Deployment is not suitable for converting to {{"apps" "__internal"} [{"apps" ""} {"apps" ""}]} in scheme "k8s.io/kubernetes/pkg/api/legacyscheme/scheme.go:29"
I0129 10:42:56.916515   76283 reflector.go:170] Listing and watching *apps.Deployment from storage/cacher.go:/deployments
I0129 10:42:56.916515   76283 reflector.go:170] Listing and watching *apps.Deployment from storage/cacher.go:/deployments
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

/sig api-machinery
/cc @caesarxuchao @deads2k 
/priority critical-urgent

### PR DESCRIPTION
**Pull request by liggitt**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Avoids artificially limiting the source API groups we support when decoding watch events from storage. The current group version selection did not handle decoding watch events from storage resulting in writes from skewed API servers across group boundaries for cohabitating resources.

**Which issue(s) this PR fixes**:
Fixes #73478

**Special notes for your reviewer**:
We'll want to take this back to all release branches. The problem exists as far back as 1.6.

**Does this PR introduce a user-facing change?**:
```release-note
fixes an error processing watch events when running skewed apiservers
```

Test output:
```
    --- PASS: TestCrossGroupStorage/apps/v1,_Kind=DaemonSet (0.06s)
        etcd_cross_group_test.go:141: wrote extensions/v1beta1 to etcd
        etcd_cross_group_test.go:163:      received event for apps/v1
        etcd_cross_group_test.go:163:      received event for apps/v1beta2
        etcd_cross_group_test.go:163:      received event for extensions/v1beta1
        etcd_cross_group_test.go:181:      fetched object for extensions/v1beta1
        etcd_cross_group_test.go:181:      fetched object for apps/v1
        etcd_cross_group_test.go:181:      fetched object for apps/v1beta2
        etcd_cross_group_test.go:141: wrote apps/v1 to etcd
        etcd_cross_group_test.go:163:      received event for extensions/v1beta1
        etcd_cross_group_test.go:163:      received event for apps/v1
        etcd_cross_group_test.go:163:      received event for apps/v1beta2
        etcd_cross_group_test.go:181:      fetched object for extensions/v1beta1
        etcd_cross_group_test.go:181:      fetched object for apps/v1
        etcd_cross_group_test.go:181:      fetched object for apps/v1beta2
        etcd_cross_group_test.go:141: wrote apps/v1beta2 to etcd
        etcd_cross_group_test.go:163:      received event for extensions/v1beta1
        etcd_cross_group_test.go:163:      received event for apps/v1
        etcd_cross_group_test.go:163:      received event for apps/v1beta2
        etcd_cross_group_test.go:181:      fetched object for extensions/v1beta1
        etcd_cross_group_test.go:181:      fetched object for apps/v1
        etcd_cross_group_test.go:181:      fetched object for apps/v1beta2
    --- PASS: TestCrossGroupStorage/networking.k8s.io/v1,_Kind=NetworkPolicy (0.03s)
        etcd_cross_group_test.go:141: wrote extensions/v1beta1 to etcd
        etcd_cross_group_test.go:163:      received event for extensions/v1beta1
        etcd_cross_group_test.go:163:      received event for networking.k8s.io/v1
        etcd_cross_group_test.go:181:      fetched object for extensions/v1beta1
        etcd_cross_group_test.go:181:      fetched object for networking.k8s.io/v1
        etcd_cross_group_test.go:141: wrote networking.k8s.io/v1 to etcd
        etcd_cross_group_test.go:163:      received event for extensions/v1beta1
        etcd_cross_group_test.go:163:      received event for networking.k8s.io/v1
        etcd_cross_group_test.go:181:      fetched object for extensions/v1beta1
        etcd_cross_group_test.go:181:      fetched object for networking.k8s.io/v1
    --- PASS: TestCrossGroupStorage//v1,_Kind=Event (0.04s)
        etcd_cross_group_test.go:141: wrote v1 to etcd
        etcd_cross_group_test.go:163:      received event for v1
        etcd_cross_group_test.go:163:      received event for events.k8s.io/v1beta1
        etcd_cross_group_test.go:181:      fetched object for v1
        etcd_cross_group_test.go:181:      fetched object for events.k8s.io/v1beta1
        etcd_cross_group_test.go:141: wrote events.k8s.io/v1beta1 to etcd
        etcd_cross_group_test.go:163:      received event for v1
        etcd_cross_group_test.go:163:      received event for events.k8s.io/v1beta1
        etcd_cross_group_test.go:181:      fetched object for v1
        etcd_cross_group_test.go:181:      fetched object for events.k8s.io/v1beta1
    --- PASS: TestCrossGroupStorage/apps/v1,_Kind=Deployment (0.06s)
        etcd_cross_group_test.go:141: wrote extensions/v1beta1 to etcd
        etcd_cross_group_test.go:163:      received event for apps/v1
        etcd_cross_group_test.go:163:      received event for apps/v1beta2
        etcd_cross_group_test.go:163:      received event for apps/v1beta1
        etcd_cross_group_test.go:163:      received event for extensions/v1beta1
        etcd_cross_group_test.go:181:      fetched object for apps/v1beta2
        etcd_cross_group_test.go:181:      fetched object for apps/v1beta1
        etcd_cross_group_test.go:181:      fetched object for extensions/v1beta1
        etcd_cross_group_test.go:181:      fetched object for apps/v1
        etcd_cross_group_test.go:141: wrote apps/v1 to etcd
        etcd_cross_group_test.go:163:      received event for apps/v1
        etcd_cross_group_test.go:163:      received event for apps/v1beta2
        etcd_cross_group_test.go:163:      received event for apps/v1beta1
        etcd_cross_group_test.go:163:      received event for extensions/v1beta1
        etcd_cross_group_test.go:181:      fetched object for extensions/v1beta1
        etcd_cross_group_test.go:181:      fetched object for apps/v1
        etcd_cross_group_test.go:181:      fetched object for apps/v1beta2
        etcd_cross_group_test.go:181:      fetched object for apps/v1beta1
        etcd_cross_group_test.go:141: wrote apps/v1beta2 to etcd
        etcd_cross_group_test.go:163:      received event for extensions/v1beta1
        etcd_cross_group_test.go:163:      received event for apps/v1
        etcd_cross_group_test.go:163:      received event for apps/v1beta2
        etcd_cross_group_test.go:163:      received event for apps/v1beta1
        etcd_cross_group_test.go:181:      fetched object for extensions/v1beta1
        etcd_cross_group_test.go:181:      fetched object for apps/v1
        etcd_cross_group_test.go:181:      fetched object for apps/v1beta2
        etcd_cross_group_test.go:181:      fetched object for apps/v1beta1
        etcd_cross_group_test.go:141: wrote apps/v1beta1 to etcd
        etcd_cross_group_test.go:163:      received event for extensions/v1beta1
        etcd_cross_group_test.go:163:      received event for apps/v1
        etcd_cross_group_test.go:163:      received event for apps/v1beta2
        etcd_cross_group_test.go:163:      received event for apps/v1beta1
        etcd_cross_group_test.go:181:      fetched object for apps/v1
        etcd_cross_group_test.go:181:      fetched object for apps/v1beta2
        etcd_cross_group_test.go:181:      fetched object for apps/v1beta1
        etcd_cross_group_test.go:181:      fetched object for extensions/v1beta1
    --- PASS: TestCrossGroupStorage/apps/v1,_Kind=ReplicaSet (0.05s)
        etcd_cross_group_test.go:141: wrote extensions/v1beta1 to etcd
        etcd_cross_group_test.go:163:      received event for extensions/v1beta1
        etcd_cross_group_test.go:163:      received event for apps/v1
        etcd_cross_group_test.go:163:      received event for apps/v1beta2
        etcd_cross_group_test.go:181:      fetched object for extensions/v1beta1
        etcd_cross_group_test.go:181:      fetched object for apps/v1
        etcd_cross_group_test.go:181:      fetched object for apps/v1beta2
        etcd_cross_group_test.go:141: wrote apps/v1 to etcd
        etcd_cross_group_test.go:163:      received event for apps/v1
        etcd_cross_group_test.go:163:      received event for apps/v1beta2
        etcd_cross_group_test.go:163:      received event for extensions/v1beta1
        etcd_cross_group_test.go:181:      fetched object for extensions/v1beta1
        etcd_cross_group_test.go:181:      fetched object for apps/v1
        etcd_cross_group_test.go:181:      fetched object for apps/v1beta2
        etcd_cross_group_test.go:141: wrote apps/v1beta2 to etcd
        etcd_cross_group_test.go:163:      received event for extensions/v1beta1
        etcd_cross_group_test.go:163:      received event for apps/v1
        etcd_cross_group_test.go:163:      received event for apps/v1beta2
        etcd_cross_group_test.go:181:      fetched object for extensions/v1beta1
        etcd_cross_group_test.go:181:      fetched object for apps/v1
        etcd_cross_group_test.go:181:      fetched object for apps/v1beta2
    --- PASS: TestCrossGroupStorage/policy/v1beta1,_Kind=PodSecurityPolicy (0.03s)
        etcd_cross_group_test.go:141: wrote extensions/v1beta1 to etcd
        etcd_cross_group_test.go:163:      received event for extensions/v1beta1
        etcd_cross_group_test.go:163:      received event for policy/v1beta1
        etcd_cross_group_test.go:181:      fetched object for extensions/v1beta1
        etcd_cross_group_test.go:181:      fetched object for policy/v1beta1
        etcd_cross_group_test.go:141: wrote policy/v1beta1 to etcd
        etcd_cross_group_test.go:163:      received event for extensions/v1beta1
        etcd_cross_group_test.go:163:      received event for policy/v1beta1
        etcd_cross_group_test.go:181:      fetched object for extensions/v1beta1
        etcd_cross_group_test.go:181:      fetched object for policy/v1beta1
```

### CODE REVIEW

**Comment by liggitt** [comment]
added a test to ensure all cross-group resources are watchable and fetchable from all exposed group/versions, no matter which version gets persisted in etcd

File: staging/src/k8s.io/apimachinery/pkg/runtime/codec.go [hunk 1]
```diff
@@ -294,6 +295,22 @@ func NewMultiGroupVersioner(gv schema.GroupVersion, groupKinds ...schema.GroupKi
...
+// Examples:
+//   gv=mygroup/__internal, groupKinds=mygroup/Foo, anothergroup/Bar
+//   KindForGroupVersionKinds(yetanother/v1/Baz, anothergroup/v1/Bar) -> mygroup/__internal/Bar (matched preferred group/kind)
+//
+//   gv=mygroup/__internal, groupKinds=mygroup, anothergroup
+//   KindForGroupVersionKinds(yetanother/v1/Baz, anothergroup/v1/Bar) -> mygroup/__internal/Bar (matched preferred group)
+//
+//   gv=mygroup/__internal, groupKinds=mygroup, anothergroup
+//   KindForGroupVersionKinds(yetanother/v1/Baz, yetanother/v1/Bar) -> mygroup/__internal/Baz (no preferred group/kind match, uses first kind in list)
+func NewCoercingMultiGroupVersioner(gv schema.GroupVersion, groupKinds ...schema.GroupKind) GroupVersioner {
```
Comment (caesarxuchao) [inline staging/src/k8s.io/apimachinery/pkg/runtime/codec.go]: A more targeted way to solve the issue is explicitly pass the cohabitating groups as the input when creating the mutleGroupVersioner. We can plumb the cohabitating groups when creating the storage code config [here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory.go#L263-L278).

Probably the benefit isn't worth the complexity. After all, if the CoercingMultiGroupVersioner returns an invalid gvk, the codec will just fail later.

**Review by caesarxuchao — COMMENTED** [review]
I can think of a more targeted fix, but I think it isn't worth the complexity.

Some minor comments.

File: test/integration/etcd/etcd_cross_group_test.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	"k8s.io/client-go/dynamic"
+	"k8s.io/kubernetes/cmd/kube-apiserver/app/options"
+)
+
+// TestCrossGroupStorage tests to make sure that all objects stored in an expected location in etcd can be converted/read.
+func TestCrossGroupStorage(t *testing.T) {
+	master := StartRealMasterOrDie(t, func(opts *options.ServerRunOptions) {
+		// force enable all resources so we can check storage.
+		// TODO: drop these once we stop allowing them to be served.
+		opts.APIEnablement.RuntimeConfig["extensions/v1beta1/deployments"] = "true"
```
[the remark this answers is not in the document]
Comment (liggitt) [reply test/integration/etcd/etcd_cross_group_test.go]: api/all enables all groups, not individual resources. this is in prep for https://github.com/kubernetes/kubernetes/pull/70672 which will disable the deprecated extensions/v1beta1 resources by default and require them to be explicitly re-enabled.

File: staging/src/k8s.io/apimachinery/pkg/runtime/codec.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/runtime/codec.go]: that was what I started with, but I actually want to give every bit of data we get back a shot at getting converted into the target version... failing early seemed artificially limiting and I couldn't think of a good reason to do that.

File: staging/src/k8s.io/apimachinery/pkg/runtime/codec.go [hunk 3]
```diff
@@ -294,6 +295,22 @@ func NewMultiGroupVersioner(gv schema.GroupVersion, groupKinds ...schema.GroupKi
 	return multiGroupVersioner{target: gv, acceptedGroupKinds: groupKinds}
 }
 
+// NewCoercingMultiGroupVersioner returns the provided group version for any incoming kind.
+// Incoming kinds that match the provided groupKinds are preferred.
+// Kind may be empty in the provided group kind, in which case any kind will match.
+// Examples:
```
[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/runtime/codec.go]: done

**Pull request merged**